firefox_puppeteer/ui/toolbars.py

```diff
@@ -0,0 +1,337 @@
+
+    @property
+    def home_button(self):
+        """
+        :returns: The home button element.
+        """
+        return self.marionette.find_element('id', 'home-button')
+
+    @property
+    def menu_button(self):
+        """
+        :returns: The menu button element.
+        """
+        return self.marionette.find_element('id', 'PanelUI-menu-button')
+
+
+class LocationBar(BaseLib):
+    """Various utilities for interacting with the location bar (the text area
+    of the ui that typically displays the current url).
+    """
+
+    dtds = ["chrome://branding/locale/brand.dtd",
+            "chrome://browser/locale/browser.dtd"]
+
+    def __init__(self, *args, **kwargs):
+        BaseLib.__init__(self, *args, **kwargs)
+        # TODO: A "utility" module that sets up the client directly would be
+        # useful here.
+        self.l10n = L10n(self.get_marionette)
+        self.keys = Keys(self.get_marionette)
+
+    @use_class_as_property('ui.toolbars.AutocompleteResults')
+    def autocomplete_results(self):
+        """Provides utility members for accessing and manipulationg the
+        locationbar.
+
+        See the :class:`~ui.toolbars.AutocompleteResults` reference.
+        """
+
+    @use_class_as_property('ui.toolbars.IdentityPopup')
+    def identity_popup(self):
+        """Provides utility members for accessing and manipulationg the
+        locationbar.
+
+        See the :class:`~ui.toolbars.IdentityPopup` reference.
+        """
+
+    @property
+    def urlbar(self):
+        """:returns: The urlbar element.
+        """
+        return self.marionette.find_element(By.ID, 'urlbar')
+
+    @property
+    def reload_button(self):
+        return self.marionette.find_element(By.ID, 'urlbar-reload-button')
+
+    @property
+    def favicon(self):
+        return self.marionette.find_element(By.ID, 'page-proxy-favicon')
+
+    @property
+    def stop_button(self):
+        return self.marionette.find_element(By.ID, 'urlbar-stop-button')
+
+    @property
+    def urlbar_input(self):
+        """:returns: The urlbar_input element
+        """
+        return self.urlbar.find_element('anon attribute', {'anonid': 'input'})
+
+    @property
+    def history_drop_marker(self):
+        return self.urlbar.find_element('anon attribute', {'anonid': 'historydropmarker'})
+
+    @property
+    def contextmenu(self):
+        """:returns: The urlbar contextmenu element.
+        """
+        parent = self.urlbar.find_element('anon attribute', {'anonid': 'textbox-input-box'})
+        return parent.find_element('anon attribute', {'anonid': 'input-box-contextmenu'})
+
+    @property
+    def notification_popup(self):
+        return self.marionette.find_element(By.ID, "notification-popup")
+
+    @property
+    def value(self):
+        """:returns: The currently displayed value of the urlbar.
+        """
+        return self.urlbar.get_attribute('value')
+
+    def get_contextmenu_entry(self, prop, subtype):
```

Comment:
I don't think we need the prop parameter here. It will always be `cmd`. `subtype` I would better rename to action. Maybe we can add a @see link to the appropriate xul file? Otherwise a list of possible actions would be good to know.